lib/elixir/lib/module.ex

```diff
@@ -1412,56 +1409,115 @@ defmodule Module do
   end
 
-  defp impl_warning(pair, kind, value, [], _callbacks) do
-    "got \"@impl #{inspect(value)}\" for #{format_definition(kind, pair)} but no behaviour was declared"
+  defp impl_behaviours(_, kind, _, _, _) when kind in [:defp, :defmacrop] do
+    {:error, :private_function}
   end
 
-  defp impl_warning(pair, kind, false, _behaviours, callbacks) do
-    case Map.fetch(callbacks, pair) do
-      {:ok, {_, behaviour, _}} ->
-        "got \"@impl false\" for #{format_definition(kind, pair)} " <>
-          "but it is a callback specified in #{inspect(behaviour)}"
+  defp impl_behaviours(_, _, value, [], _) do
+    {:error, {:no_behaviours, value}}
+  end
 
-      :error ->
-        nil
+  defp impl_behaviours(impls, _, false, _, callbacks) do
+    case impl_callbacks(impls, callbacks) do
+      [] -> {:ok, []}
+      [impl | _] -> {:error, {:impl_not_defined, impl}}
     end
   end
 
-  defp impl_warning(pair, kind, true, _, callbacks) do
-    if not Map.has_key?(callbacks, pair) do
-      "got \"@impl true\" for #{format_definition(kind, pair)} " <>
-        "but no behaviour specifies such callback#{known_callbacks(callbacks)}"
+  defp impl_behaviours(impls, _, true, _, callbacks) do
+    case impl_callbacks(impls, callbacks) do
+      [] -> {:error, {:impl_defined, callbacks}}
+      impls -> {:ok, impls}
     end
   end
 
-  defp impl_warning(pair, kind, behaviour, behaviours, callbacks) do
+  defp impl_behaviours(impls, _, behaviour, behaviours, callbacks) do
+    filtered = impl_behaviours(impls, behaviour, callbacks)
+
     cond do
-      match?({:ok, {_, ^behaviour, _}}, Map.fetch(callbacks, pair)) ->
-        nil
+      filtered != [] ->
+        {:ok, filtered}
 
       behaviour not in behaviours ->
-        "got \"@impl #{inspect(behaviour)}\" for #{format_definition(kind, pair)} " <>
-          "but this behaviour was not declared with @behaviour"
+        {:error, {:behaviour_not_declared, behaviour}}
 
       true ->
-        "got \"@impl #{inspect(behaviour)}\" for #{format_definition(kind, pair)} " <>
-          "but this behaviour does not specify such callback#{known_callbacks(callbacks)}"
+        {:error, {:behaviour_not_defined, behaviour, callbacks}}
     end
   end
 
+  defp impl_behaviours(impls, behaviour, callbacks) do
+    impl_behaviours(impl_callbacks(impls, callbacks), behaviour)
+  end
+
+  defp impl_behaviours([], _) do
+    []
+  end
+
+  defp impl_behaviours([{_, behaviour} = impl | tail], behaviour) do
+    [impl | impl_behaviours(tail, behaviour)]
+  end
+
+  defp impl_behaviours([_ | tail], behaviour) do
+    impl_behaviours(tail, behaviour)
+  end
+
+  defp impl_callbacks([], _) do
+    []
+  end
+
+  defp impl_callbacks([fa | tail], callbacks) do
+    case callbacks[fa] do
+      nil -> impl_callbacks(tail, callbacks)
+      {_, behaviour, _} -> [{fa, behaviour} | impl_callbacks(tail, callbacks)]
+    end
+  end
+
+  def format_message(fa, kind, :private_function) do
```

Comment:
Let's name this to "format_impl_warning" and make sure it is private. :)

Reply:
Done c00c155! Good catch on the private, I missed that